torch/distributed/rpc/api.py

```diff
@@ -331,9 +332,20 @@ def shutdown(graceful=True):
         >>> rpc.shutdown()
     """
     if graceful:
-        _wait_all_workers()
+        try:
+            _wait_all_workers()
+        except RuntimeError as ex:
+            # Following steps are bound to fail, so just finalize
+            # shutdown and exit.
+            _finalize_shutdown()
+            return
+
         _delete_all_user_and_unforked_owner_rrefs()
         _get_current_rpc_agent().join(shutdown=True)
+
+    _finalize_shutdown()
+
+def _finalize_shutdown():
```

Comment:
Nit: this seems a nice use-case for the little know `try-else` syntax:
```suggestion
        try:
            _wait_all_workers()
        else:
            _delete_all_user_and_unforked_owner_rrefs()
            _get_current_rpc_agent().join(shutdown=True)
```

Reply:
My understanding is that else clause will run only if there are no exceptions. But in case there is an exception, we still need to catch it and call `finalize_shutdown` right? Otherwise the exception from `_wait_all_workers` will just be propagated up the stack without shutting down the RPC framework.

Reply:
Can't this be achieved with a try-else-finally clause?

Reply:
I guess try-finally should be sufficient?

Reply:
Yeah probably. Possibly try-else-finally could be "clearer" as it shows which of the three statements we expect could raise an exception. Up to you.